### PR TITLE
Convert tests to pytest (assorted)

### PR DESCRIPTION
**Pull request by jakelishman**
These are all the remaining files from #1181 that neither Eric nor Boxi had explicitly said they were checking, with the commits all tidied up and rebased onto `master`.

This is based on #1249, and that one needs to be merged first.

**Changelog**
Major test refactor into a pytest style.

### CODE REVIEW

File: qutip/tests/conftest.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
+# This file is part of QuTiP: Quantum Toolbox in Python.
+#
+#    Copyright (c) 2011 and later, Paul D. Nation and Robert J. Johansson.
+#    All rights reserved.
+#
+#    Redistribution and use in source and binary forms, with or without
+#    modification, are permitted provided that the following conditions are
+#    met:
+#
```
Comment (nathanshammah) [inline qutip/tests/conftest.py:9]:
```suggestio
#    modification, are permitted provided that the following conditions ar
#    met
#
```

File: qutip/tests/conftest.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+#    2. Redistributions in binary form must reproduce the above copyright
+#       notice, this list of conditions and the following disclaimer in the
+#       documentation and/or other materials provided with the distribution.
+#
+#    3. Neither the name of the QuTiP: Quantum Toolbox in Python nor the names
+#       of its contributors may be used to endorse or promote products derived
+#       from this software without specific prior written permission.
+#
+#    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+#    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
```
Comment (nathanshammah) [inline qutip/tests/conftest.py:22]:
```suggestion
#
#    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
#    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
```
```suggestion
#
#    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
#    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
```

File: qutip/tests/test_entropy.py [hunk 3]
```diff
@@ -31,117 +31,93 @@
...
+        pytest.param(qutip.bell_state(x).proj(), id='bell'+x)
+        for x in ['00', '01', '10', '11']
+    ])
+    def test_maximally_entangled(self, dm):
+        assert abs(qutip.concurrence(dm) - 1) < 1e-12
+
+    @pytest.mark.repeat(10)
+    def test_nonzero(self):
+        dm = qutip.rand_dm(4, dims=[[2, 2], [2, 2]])
+        assert qutip.concurrence(dm) >= 0
```
Comment (nathanshammah) [inline qutip/tests/test_entropy.py:73]:
```suggestion
    @pytest.mark.repeat(11)
    def test_nonzero(self):
        dm = qutip.rand_dm(4, dims=[[2, 2], [2, 2]])
        assert qutip.concurrence(dm) >= 0
```

**Comment by jakelishman** [comment]
@nathanshammah are you just trying out the suggestions feature, or are these actual changes to be made?

**Comment by BoxiLi** [comment]
He was just trying out. I think he was demonstrating this in one of our meetings, lol. Sorry for the confusion :)

**Comment by nathanshammah** [comment]
Yes sorry

**Pull request merged**